doc/barefoot_dtel/Dtel-SONiC.md

```diff
@@ -0,0 +1,509 @@
+INNER_IP_PROTOCOL           = 1*2HEXDIG ; IP protocol type of the inner header
+INNER_L4_SRC_PORT           = 1*5DIGIT  ; a number between 0 and 65535
+INNER_L4_DST_PORT           = 1*5DIGIT  ; a number between 0 and 65535
+FLOW_OP                     = "NOP" / "POSTCARD" / "INT" / "IOAM" ; Applicable only when 
+                                                                  ; table name is DTEL_FLOW_WATCHLIST
+INT_SESSION                 = 1*255VCHAR ; previously configured INT-session-name
+                                         ; Applicable only when FLOW_OP = INT or IOAM
+DROP_REPORT_ENABLE          = "TRUE" / "FALSE" ; Applicable only when table name is DTEL_DROP_WATCHLIST
+                                               ; Note: FLOW_OP is not set when this is set
+REPORT_TAIL_DROPS           = "TRUE" / "FALSE" ; Applicable only when table name is DTEL_DROP_WATCHLIST
+                                               ; and DROP_REPORT_ENABLE = "TRUE"
+FLOW_SAMPLE_PERCENT         = 1*DIGIT ; number between 0 to 100
+                                      ; Applicable only when FLOW_OP = INT or IOAM or POSTCARD
+REPORT_ALL_PACKETS          = "TRUE" / "FALSE" ; Applicable only when FLOW_OP = INT or IOAM or POSTCARD
+
+;value annotations
+table_name  =  "DTEL_FLOW_WATCHLIST" / "DTEL_DROP_WATCHLIST" ; 
+
+```
+
+Example configuration using redis-cli
+
+    HMSET ACL_RULE|DTEL_FLOW_WATCHLIST|FW-1 PRIORITY 10 SRC_IP 10.1.1.1 DST_IP 20.1.1.1 L4_SRC_PORT 1000 L4_DST_PORT 2000 INNER_ETHER_TYPE 55 INNER_L4_SRC_PORT 3000 INNER_L4_DST_PORT 4000 FLOW_OP POSTCARD FLOW_SAMPLE_PERCENT 80 REPORT_ALL_PACKETS TRUE
+    HMSET ACL_RULE|DTEL_FLOW_WATCHLIST|FW-2 PRIORITY 10 INT_SESSION INT-1 FLOW_OP INT
+    HSET ACL_RULE|DTEL_DROP_WATCHLIST|DW-1 PRIORITY 10 L4_DST_PORT 2000 DROP_REPORT_ENABLE TRUE
+    
+
+## DTel orchagent
+
+This is a new Orch agent added to handle DTel events. These events are demuxed based on the ConfigDB table name.
+
+Events handled by DtelOrch agent:
+
+    * Set event
+    * Delete event
+
+Most events are handled as shown in the sequence diagram below. Ones which are different are depicted in the following figures.
+
+![Control flow](cflow-1.jpg)
+
+__Figure 2: Generic control flow for DTel events__. 
+
+  
+![Control flow](cflow-2.jpg)
+
+__Figure 3: Control flow for DTel events that depend on other Orch agents__. 
+
+
+### Ref-counted DTel objects
+
+These objects cannot be deleted unless their ref-counts are zero.
+
+* INT sessions referenced by INT watchlists
+* Report sessions referenced by DTel events
+
+![Control flow](cflow-3.jpg "Control flow")
+
+__Figure 4: Control flow for DTel events corresponding to ref-counted objects__. 
+
+  
+### State maintained by DTelOrch agent
+
+**INT session hashmap**
+
+* Key: INT session id
+* Value: 
+	* INT session OID
+	* Reference count
+
+**Report session hashmap**
+
+* Key: Report session ID
+* Value: 
+	* Report session OID
+	* Reference count
+
```

Comment:
I want to follow up with the reference count discussion we had during the meeting:

In meeting, the reference count usage was described as:
1. When reference count is not zero, the object is protected from deletion.
2. When reference count drops to zero, a delete request will free the object.

I have concern for this approach. The danger is that a delete call at the right time is required to free the object. So the caller of delete will have to check if the object has been freed or denied of deletion. This makes code logic complicated and encourages opportunities for either memory leak or double free.

I think a safe approach to use reference count is following:
- An object get reference count of 1 at creation time.
- A delete request reduces the reference count by 1.
- An association with another entity increases the reference count by 1.
- Detaching from another entity decrease the reference count by 1.
- Object is freed when the reference count drops to 0.

The advantage of this approach is that the delete can be called whenever the code decide to free. And the object will be freed when all the associations are gone. This makes code logical simpler.